### DISCUSSION TITLE
How do you "reduce" or "fold" a list?

### DISCUSSION THREAD

**Discussion by asthasr** [cozodb/cozo#129] [Q&A]
If I call `sum(...)` on a list of numbers, I receive the error `  × cannot compute 'sum': encountered value [1, 1, 1]`. `add` gives `help: addition requires numbers`. Is there a function or operator I'm overlooking?

**Comment by infogulch** [comment]
```
{ # tx 1: create 4 rows with one column each, place them into a new relation named "nums"
?[n] <- [[1],[2],[3],[4]]
:create nums{n}
}
{ # tx 2: query nums relation and sum them
?[sum(n)] := *nums{n}
}
```

**Reply by asthasr** [reply]
That's not exactly what I mean. If you have a relation with a column that consists of a list, e.g.

```
name, scores
alice, [95, 90, 90, 93, 100]
bob, [88, 91, 100, 97, 85]
```

How would I apply a folding function when querying that column?

**Comment by gisborne** [comment]
The answer is something like this:

```
{
    ?[n, s] <- [['alice', [95, 90, 90, 93, 100]],
                ['bob', [88, 91, 100, 97, 85]]]

:create folk{n, s}
}
{
    has_val_list[l, x] := x = first(l)
    has_val_list[l, x] := gt(length(l), 1), has_val_list[slice(l, 1, -1), x]
    ?[name, val] := *folk[name, s], has_val_list[s, val]
}
```

This doesn't actually work — Cozo complains about about an unbound variable in `has_val_list[l, x] := x = first(l)`.

I just realised there is no real documentation on how to use functions.

Anyone know why this doesn't work?

**Reply by jr-acosta** [reply]
{
?[n, s] <- [['alice', [95, 90, 90, 93, 100]],['bob', [88, 91, 100, 97, 85]]]

:replace folk{n, s}
}
{

r[t,l,n] := *folk[n,l],t=0
r[t,w,n] := *folk[n,underscore1],gt(length(l), 0),r[underscore2,l,n], w=slice(l,0,-1), t= last(l)
?[n,sum(scores)] := *folk[n,underscore3],r[scores,underscore4,n]
}


Run this code and see if it works for you.

**Comment by tom-bywild** [comment]
I think for that feature, you have to define a custom named rule, if supported by your env. E.g. the [Node-example](https://github.com/cozodb/cozo/blob/main/cozo-lib-nodejs/example.js#L26) implements a custom `Pipipy`.

I just started to learn CozoScript and its design & and approach to querying, which is very new for me and super exciting. AFAIK, the aggregations only apply on a _row_-level for results.

I guess the `add`-function, as you mentioned, would be the right choice, but there's currently no way to cast lists to variadics. So I guess all arithmetic ops would be great candidates to support lists as well.

**Reply by gisborne** [reply]
Why does my example not work?

**Reply by jr-acosta** [reply]
{
?[n, s] <- [['alice', [95, 90, 90, 93, 100]],['bob', [88, 91, 100, 97, 85]]]

:replace folk{n, s}
}
{

r[t,l,n] := *folk[n,l],t=0
r[t,w,n] := *folk[n,underscore1],gt(length(l), 0),r[underscore2,l,n], w=slice(l,0,-1), t= last(l)
 ?[n,sum(scores)] := *folk[n,underscore3],r[scores,underscore4,n]
}


Run this code and see if it works for you.
